File: integrasjonspunkt/src/main/java/no/difi/meldingsutveksling/nextmove/v2/NextMoveFileSizeValidator.java
package no.difi.meldingsutveksling.nextmove.v2;

import no.difi.meldingsutveksling.ServiceIdentifier;
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties;
import no.difi.meldingsutveksling.exceptions.MaxFileSizeExceededException;
import no.difi.meldingsutveksling.nextmove.NextMoveOutMessage;
import no.difi.meldingsutveksling.nextmove.NextMoveRuntimeException;
import org.springframework.stereotype.Component;
import org.springframework.util.unit.DataSize;
import org.springframework.web.multipart.MultipartFile;

@Component
public class NextMoveFileSizeValidator {

    private final IntegrasjonspunktProperties props;

    public NextMoveFileSizeValidator(IntegrasjonspunktProperties props) {
        this.props = props;
    }

    public void validate(NextMoveOutMessage msg, MultipartFile file) {
        long totalBytes = file.getSize();
        if (msg.getFiles() != null) {
            for (var f : msg.getFiles()) {
                totalBytes += f.getSize();
            }
        }
        DataSize total = DataSize.ofBytes(totalBytes);
        DataSize limit;
        ServiceIdentifier si = msg.getServiceIdentifier();
        switch (si) {
            case DPO:
                limit = props.getDpo().getUploadSizeLimit();
                break;
            case DPV:
                limit = props.getDpv().getUploadSizeLimit();
                break;
            case DPE:
                limit = props.getNextmove().getServiceBus().getUploadSizeLimit();
                break;
            case DPF:
                limit = props.getFiks().getUt().getUploadSizeLimit();
                break;
            case DPI:
                limit = props.getDpi().getUploadSizeLimit();
                break;
            case DPFIO:
                limit = props.getFiks().getIo().getUploadSizeLimit();
                break;
            case DPH:
                limit = props.getDph().getUploadSizeLimit();
                break;
            default:
                throw new NextMoveRuntimeException("Unknown Service Identifier");
        }
        if (total.compareTo(limit) > 0) {
            if (total.toBytes() < DataSize.ofMegabytes(1L).toBytes()) {
                throw new MaxFileSizeExceededException(total.toBytes() + "b", si.toString(), limit.toBytes() + "b");
            }
            throw new MaxFileSizeExceededException(total.toMegabytes() + "MB", si.toString(), limit.toMegabytes() + "MB");
        }
    }

}

File: integrasjonspunkt/src/main/kotlin/no/difi/meldingsutveksling/nextmove/v2/NextMoveFileSizeValidator.kt
package no.difi.meldingsutveksling.nextmove.v2

import no.difi.meldingsutveksling.ServiceIdentifier.*
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties
import no.difi.meldingsutveksling.exceptions.MaxFileSizeExceededException
import no.difi.meldingsutveksling.nextmove.NextMoveOutMessage
import no.difi.meldingsutveksling.nextmove.NextMoveRuntimeException
import org.springframework.stereotype.Component
import org.springframework.util.unit.DataSize
import org.springframework.web.multipart.MultipartFile

@Component
class NextMoveFileSizeValidator(private val props: IntegrasjonspunktProperties) {

    fun validate(msg: NextMoveOutMessage, file: MultipartFile) {
        val total = DataSize.ofBytes(msg.files.map { it.size }.fold(file.size) { a, b -> a + b })
        val limit = when (msg.serviceIdentifier) {
            DPO -> props.dpo.uploadSizeLimit
            DPV -> props.dpv.uploadSizeLimit
            DPE -> props.nextmove.serviceBus.uploadSizeLimit
            DPF -> props.fiks.ut.uploadSizeLimit
            DPI -> props.dpi.uploadSizeLimit
            DPFIO -> props.fiks.io.uploadSizeLimit
            DPH -> props.dph.uploadSizeLimit
            else -> throw NextMoveRuntimeException("Unknown Service Identifier")
        }
        if (total > limit) {
            if (total < DataSize.ofMegabytes(1L)) {
                throw MaxFileSizeExceededException("${total.toBytes()}b", msg.serviceIdentifier.toString(), "${limit.toBytes()}b")
            }
            throw MaxFileSizeExceededException("${total.toMegabytes()}MB", msg.serviceIdentifier.toString(), "${limit.toMegabytes()}MB")
        }
    }
}

